python_modules/dagster/dagster/_daemon/monitoring/monitoring_daemon.py

```diff
@@ -137,12 +178,18 @@ def execute_monitoring_iteration(
         try:
             logger.info(f"Checking run {run_record.dagster_run.run_id}")
 
-            if run_record.dagster_run.status == DagsterRunStatus.STARTING:
+            if (
+                instance.run_monitoring_start_timeout_seconds > 0
+                and run_record.dagster_run.status == DagsterRunStatus.STARTING
+            ):
                 monitor_starting_run(instance, run_record, logger)
             elif run_record.dagster_run.status == DagsterRunStatus.STARTED:
                 monitor_started_run(instance, workspace, run_record, logger)
-            elif run_record.dagster_run.status == DagsterRunStatus.CANCELING:
-                # TODO: implement canceling timeouts
+            elif (
+                instance.run_monitoring_cancel_timeout_seconds > 0
+                and run_record.dagster_run.status == DagsterRunStatus.CANCELING
+            ):
+                monitor_canceling_run(instance, run_record, logger)
```

Comment:
Seems we could restructure this to not recalculate "when X entered state Y" every tick and only when its first observed. Looks like we suggest a high poll interval so not clearly a high priority.